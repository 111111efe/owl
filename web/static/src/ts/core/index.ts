export { QWeb } from "./qweb_vdom";
export { EventBus } from "./event_bus";
export { Component } from "./component";
