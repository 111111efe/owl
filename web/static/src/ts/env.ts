import { WEnv } from "./core/component";
import { QWeb } from "./core/qweb_vdom";
import { Registry } from "./core/registry";
import { idGenerator, memoize } from "./core/utils";
import { TemplateLoader } from "./loaders/templates";
import { actionRegistry } from "./registries";
import { RPC } from "./services/ajax";
import {
  ActionManager,
  ActionWidget,
  IActionManager
} from "./store/action_manager";
import { Ajax, IAjax } from "./store/ajax";
import {
  INotificationManager,
  NotificationManager
} from "./store/notifications";
import { IRouter, Router } from "./store/router";

//------------------------------------------------------------------------------
// Types
//------------------------------------------------------------------------------

export interface Env extends WEnv {
  // services
  actionManager: IActionManager;
  ajax: IAjax;
  notifications: INotificationManager;
  router: IRouter;

  // registries
  actionRegistry: Registry<ActionWidget>;

  // helpers
  rpc: IAjax["rpc"];

  // configuration
  debug: boolean;
  isMobile: boolean;
}

interface Loaders {
  loadTemplates: TemplateLoader;
}

interface Services {
  rpc: RPC;
}

type EnvBuilder = (loaders: Loaders, services: Services) => Promise<Env>;
//------------------------------------------------------------------------------
// Environment
//------------------------------------------------------------------------------

/**
 * init returns the main environment for the application.
 *
 * Note that it does not make much sense (except for tests) to have more than
 * one environment. For example, with two environment, the router code in one
 * environment will probably interfere with the code from the other environment.
 *
 * For this reason, the result of init is memoized: every call to
 * this function will actually return the same environment.
 */
export const makeEnv: EnvBuilder = memoize(_makeEnv);

async function _makeEnv(loaders: Loaders, services: Services): Promise<Env> {
  // services
  const qweb = new QWeb();
  const router = new Router();
  const ajax = new Ajax(services.rpc);
  const actionManager = new ActionManager(actionRegistry, ajax);
  const notifications = new NotificationManager();

  // templates
  const templates = await loaders.loadTemplates();
  qweb.addTemplate("default", "<div/>");
  qweb.loadTemplates(templates);

  const env: Env = {
    // Base widget requirements
    qweb,
    getID: idGenerator(),

    actionManager,
    ajax,
    notifications,
    actionRegistry,
    router,

    rpc: ajax.rpc,

    debug: false,
    isMobile: window.innerWidth <= 768
  };
  return env;
}
