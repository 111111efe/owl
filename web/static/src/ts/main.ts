///<amd-module name="main" />

import { makeEnvironment } from "./env";
import { BaseMenuItem, processMenuItems } from "./misc/menu_helpers";
import { Root } from "./widgets/root";

//------------------------------------------------------------------------------
// Application bootstrapping
//------------------------------------------------------------------------------

document.addEventListener("DOMContentLoaded", async function() {
  const env = await makeEnvironment();

  // menu processing
  const menuItems: BaseMenuItem[] = (<any>window).odoo.menus;
  const menuInfo = processMenuItems(menuItems);
  delete (<any>window).odoo.menus; // overkill?

  // creating root widget
  const rootWidget = new Root(env, { menuInfo });
  await rootWidget.mount(document.body);
  (<any>window).odoo.rootWidget = rootWidget;
});
