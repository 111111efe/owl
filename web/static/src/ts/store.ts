import { Type } from "./core/component";
import { EventBus } from "./core/event_bus";
import { Registry } from "./core/registry";
import { RPC } from "./services/ajax";
import { IRouter, Query } from "./services/router";
import { Widget } from "./widgets/widget";

//------------------------------------------------------------------------------
// Types
//------------------------------------------------------------------------------

export interface MenuItem {
  id: number;
  name: string;
  parentId: number | false;
  action: string | false;
  icon: string | false;

  // root menu id
  app: MenuItem;
  actionId: number;
  children: MenuItem[];
}

export interface MenuInfo {
  menus: { [key: number]: MenuItem | undefined };

  actionMap: { [id: number]: MenuItem | undefined };
  roots: number[];
}

export interface State {
  stack: ActionStack;
  inHome: boolean;
  currentApp: MenuItem | null;
}

export interface Services {
  rpc: RPC;
  router: IRouter;
}

export type Context = { [key: string]: any };

export interface CommonActionInfo {
  id: number;
  context: Context;
  title: string;
  target: "current" | "new";
}

export type ActionRequest = string | number;

export type ActionWidget = Type<Widget<{}, {}>>;

export interface ClientActionInfo extends CommonActionInfo {
  type: "client";
  name: string;
  Widget: ActionWidget;
}

export interface ActWindowInfo extends CommonActionInfo {
  type: "act_window";
  view: string;
}

export interface ActionDescription {
  id: number;
  type: "ir.actions.act_window" | "ir.actions.client";
  target: "current";
}

export type ActionInfo = ClientActionInfo | ActWindowInfo;
export type ActionStack = ActionInfo[];

export interface RPCModelQuery {
  model: string;
  method: string;
  args?: any[];
  kwargs?: { [key: string]: any };
  context?: { [key: string]: any };
}

export interface RPCControllerQuery {
  route: string;
  params: { [key: string]: any };
}

export type RPCQuery = RPCModelQuery | RPCControllerQuery;

export type RPC = (rpc: RPCQuery) => Promise<any>;

export interface INotification {
  id: number;
  title: string;
  message: string;
  type: "notification" | "warning";
  sticky: boolean;
}

interface RequestParameters {
  route: string;
  params: { [key: string]: any };
}

//------------------------------------------------------------------------------
// Store
//------------------------------------------------------------------------------

export class Store extends EventBus {
  state: State;
  menuInfo: MenuInfo;
  services: Services;
  actionRegistry: Registry<ActionWidget>;

  constructor(
    services: Services,
    menuInfo: MenuInfo,
    actionRegistry: Registry<ActionWidget>
  ) {
    super();
    this.services = services;
    this.menuInfo = menuInfo;
    this.actionRegistry = actionRegistry;
    this.state = {
      stack: [],
      inHome: false,
      currentApp: null
    };
    const query = this.services.router.getQuery();
    let { app, actionId } = this.getAppAndAction(query);
    this.state.currentApp = app;
    if (!actionId) {
      this.state.inHome = true;
    }

    this.services.router.on("query_changed", this, this.updateAction);
    this.updateAction(this.services.router.getQuery());
  }

  private updateAction(query: Query) {
    let { app, actionId } = this.getAppAndAction(query);
    this.updateAppState(app, actionId);
  }

  dispatch(action: string, params?: any) {
    switch (action) {
      case "open_menu":
        this.updateAppState(params.app, params.actionId);
        break;
      case "toggle_home_menu":
        this.updateState({ inHome: !this.state.inHome });
        break;
      case "add_notification":
        this.addNotification(params);
        break;
      case "close_notification":
        this.closeNotification(params);
        break;
    }
  }

  updateState(nextState: Partial<State>) {
    Object.assign(this.state, nextState);
    this.trigger("state_updated", this.state);
  }

  nextID = 1;

  addNotification(notif: Partial<INotification>): number {
    const id = this.nextID++;
    const defaultVals = {
      title: "",
      message: "",
      type: "notification",
      sticky: false
    };
    const notification = Object.assign(defaultVals, notif, { id });
    this.trigger("notification_added", notification);
    if (!notification.sticky) {
      setTimeout(() => this.closeNotification(id), 2500);
    }
    return id;
  }
  closeNotification(id: number) {
    this.trigger("notification_closed", id);
  }

  private updateAppState(app: MenuItem | null, actionId: number | null) {
    const newApp = app || this.state.currentApp;
    if (actionId) {
      const query: Query = { action_id: String(actionId) };
      const menuId = newApp ? newApp.app.id : false;
      if (menuId) {
        query.menu_id = String(menuId);
      }
      if (app) {
        this.updateState({ currentApp: app });
      }
      this.services.router.navigate(query);
      this.doAction(actionId);
    } else {
      this.updateState({ inHome: true, currentApp: newApp });
    }
  }

  private getAppAndAction(
    query: Query
  ): { app: MenuItem | null; actionId: number | null } {
    const menuInfo = this.menuInfo;
    let app: MenuItem | null = null;
    let actionId: number | null = null;
    if ("action_id" in query) {
      actionId = parseInt(query.action_id, 10);
      if (menuInfo.actionMap[actionId]) {
        const menu = menuInfo.actionMap[actionId]!;
        app = menu.app;
      }
    }
    if ("menu_id" in query) {
      const menuId = parseInt(query.menu_id, 10);
      const menu = menuInfo.menus[menuId];
      if (menu) {
        app = menu.app;
        if (!actionId) {
          actionId = menu.actionId;
        }
      }
    }
    return { app, actionId };
  }

  doAction(request: ActionRequest) {
    if (typeof request === "number") {
      this.loadAction(request);
      // this is an action ID
      let name = request === 131 ? "discuss" : "crm";
      let title =
        request === 131 ? "Discuss" : request === 250 ? "Notes" : "CRM";
      let Widget = this.actionRegistry.get(name);
      this.updateState({
        inHome: false,
        stack: [
          {
            id: 1,
            context: {},
            target: "current",
            type: "client",
            name,
            title,
            Widget: Widget
          }
        ]
      });
    }
  }

  private loadAction(id: number) {
    this.rpc({
      route: "web/action/load",
      params: {
        action_id: id
      }
    });
  }

  counter: number = 0;

  async rpc(rpc: RPCQuery): Promise<any> {
    const request = this.prepareRequest(rpc);
    if (this.counter === 0) {
      this.trigger("rpc_status", "loading");
    }
    this.counter++;
    const result = await this.services.rpc(request.route, request.params);
    this.counter--;
    if (this.counter === 0) {
      this.trigger("rpc_status", "notloading");
    }
    return result;
  }

  private prepareRequest(query: RPCQuery): RequestParameters {
    let route: string;
    let params = "params" in query ? query.params : {};
    if ("route" in query) {
      route = query.route;
    } else if ("model" in query && "method" in query) {
      route = `/web/dataset/call_kw/${query.model}/${query.method}`;
      params.args = query.args || [];
      params.model = query.model;
      params.method = query.method;
      params.kwargs = Object.assign(params.kwargs || {}, query.kwargs);
      params.kwargs.context =
        query.context || params.context || params.kwargs.context;
    } else {
      throw new Error("Invalid Query");
    }

    // doing this remove empty keys, and undefined stuff
    const sanitizedParams = JSON.parse(JSON.stringify(params));
    return { route, params: sanitizedParams };
  }
}
