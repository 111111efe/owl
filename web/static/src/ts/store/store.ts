import { EventBus } from "../core/event_bus";
import { Registry } from "../core/registry";
import { RPC } from "../services/ajax";
import { IRouter, Query } from "../services/router";
import {
  actionManagerMixin,
  ActionStack,
  ActionWidget
} from "./action_manager_mixin";
import { notificationMixin } from "./notification_mixin";
import { rpcMixin } from "./rpc_mixin";
import { MenuItem } from "./store";

//------------------------------------------------------------------------------
// Types
//------------------------------------------------------------------------------

export { ActionStack, ActionWidget } from "./action_manager_mixin";
export { INotification } from "./notification_mixin";
export { RPC } from "./rpc_mixin";

export interface MenuItem {
  id: number;
  name: string;
  parentId: number | false;
  action: string | false;
  icon: string | false;

  // root menu id
  app: MenuItem;
  actionId: number;
  children: MenuItem[];
}

export interface MenuInfo {
  menus: { [key: number]: MenuItem | undefined };

  actionMap: { [id: number]: MenuItem | undefined };
  roots: number[];
}

export interface State {
  stack: ActionStack;
  inHome: boolean;
  currentApp: MenuItem | null;
}

export interface Services {
  rpc: RPC;
  router: IRouter;
}

//------------------------------------------------------------------------------
// Store
//------------------------------------------------------------------------------
export class BaseStore extends EventBus {
  state: State = {
    stack: [],
    inHome: false,
    currentApp: null
  };
  menuInfo: MenuInfo;
  services: Services;
  actionRegistry: Registry<ActionWidget>;
  currentQuery: Query;

  constructor(
    services: Services,
    menuInfo: MenuInfo,
    actionRegistry: Registry<ActionWidget>
  ) {
    super();
    this.services = services;
    this.menuInfo = menuInfo;
    this.actionRegistry = actionRegistry;
    this.currentQuery = {};
  }

  update(nextState: Partial<State>) {
    Object.assign(this.state, nextState);
    this.trigger("state_updated", this.state);
  }

  toggleHomeMenu() {
    if (this.state.inHome && !this.state.currentApp) {
      return;
    }
    this.update({ inHome: !this.state.inHome });
    if (this.state.inHome) {
      this.services.router.navigate({ home: true });
    } else {
      this.updateQuery(this.currentQuery);
    }
  }

  updateQuery(query: Query) {
    this.currentQuery = query;
    this.services.router.navigate(query);
  }
}

export class Store extends actionManagerMixin(
  rpcMixin(notificationMixin(BaseStore))
) {
  constructor(
    services: Services,
    menuInfo: MenuInfo,
    actionRegistry: Registry<ActionWidget>
  ) {
    super(services, menuInfo, actionRegistry);
    const query = this.services.router.getQuery();
    let { app, actionId } = this.getAppAndAction(query);
    this.state.currentApp = app;
    if (!actionId) {
      this.state.inHome = true;
      this.services.router.navigate({ home: true });
    }

    this.services.router.on("query_changed", this, this.updateAction);
    this.updateAction(this.services.router.getQuery());
  }

  private updateAction(query: Query) {
    let { app, actionId } = this.getAppAndAction(query);
    this.updateAppState(app, actionId);
  }

  activateMenuItem(menuId: number) {
    const menu = this.menuInfo.menus[menuId];
    if (!menu) {
      throw new Error("Invalid menu id");
    }
    return this.updateAppState(menu.app, menu.actionId);
  }

  private async updateAppState(app: MenuItem | null, actionId: number | null) {
    const newApp = app || this.state.currentApp;
    if (actionId) {
      const query: Query = { action_id: String(actionId) };
      const menuId = newApp ? newApp.app.id : false;
      if (menuId) {
        query.menu_id = String(menuId);
      }
      if (app) {
        this.update({ currentApp: app });
      }
      await this.doAction(actionId);
      this.updateQuery(query);
    } else {
      this.update({ inHome: true, currentApp: newApp });
    }
  }

  getAppAndAction(
    query: Query
  ): { app: MenuItem | null; actionId: number | null } {
    const menuInfo = this.menuInfo;
    let app: MenuItem | null = null;
    let actionId: number | null = null;
    if (typeof query.action_id === "string") {
      actionId = parseInt(query.action_id, 10);
      if (menuInfo.actionMap[actionId]) {
        const menu = menuInfo.actionMap[actionId]!;
        app = menu.app;
      }
    }
    if (typeof query.menu_id === "string") {
      const menuId = parseInt(query.menu_id, 10);
      const menu = menuInfo.menus[menuId];
      if (menu) {
        app = menu.app;
        if (!actionId) {
          actionId = menu.actionId;
        }
      }
    }
    return { app, actionId };
  }
}
