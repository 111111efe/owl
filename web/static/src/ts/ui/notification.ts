import { INotification } from "../store/store";
import { Widget } from "./widget";

export class Notification extends Widget<INotification, {}> {
  template = "web.notification";

  close(ev: MouseEvent) {
    // we do not want the url to change
    ev.preventDefault();
    this.env.closeNotification(this.props.id);
  }
}
