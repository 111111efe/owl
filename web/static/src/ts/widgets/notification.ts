import { INotification } from "../store";
import { Widget } from "./widget";

export class Notification extends Widget<INotification, {}> {
  template = "web.notification";

  close(ev: MouseEvent) {
    // we do not want the url to change
    ev.preventDefault();
    this.env.dispatch("close_notification", this.props.id);
  }
}
