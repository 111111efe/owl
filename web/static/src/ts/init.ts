import { Ajax } from "./store/ajax";
import { NotificationManager } from "./core/notifications";
import { QWeb } from "./core/qweb_vdom";
import { Router } from "./core/router";
import { findInTree, idGenerator, memoize } from "./core/utils";
import { actionRegistry } from "./registries";
import { ActionManager } from "./store/action_manager";
import { MenuInfo, MenuItem } from "./widgets/root";
import { Env } from "./widgets/widget";

//------------------------------------------------------------------------------
// Types
//------------------------------------------------------------------------------

interface InitializedData {
  env: Env;
  menuInfo: MenuInfo;
}

/**
 * init returns the main environment for the application.
 *
 * Note that it does not make much sense (except for tests) to have more than
 * one environment. For example, with two environment, the router code in one
 * environment will probably interfere with the code from the other environment.
 *
 * For this reason, the result of init is memoized: every call to
 * this function will actually return the same environment.
 */
export const init = memoize(async function(): Promise<InitializedData> {
  // services
  const qweb = new QWeb();
  const router = new Router();
  const ajax = new Ajax(ajaxFetch);
  const actionManager = new ActionManager(actionRegistry, ajax);
  const notifications = new NotificationManager();

  // templates
  const templates = await loadTemplates();
  qweb.addTemplate("default", "<div/>");
  qweb.loadTemplates(templates);

  const menuInfo = loadMenus();

  const env: Env = {
    // Base widget requirements
    qweb,
    getID: idGenerator(),

    actionManager,
    ajax,
    notifications,
    actionRegistry,
    router,

    rpc: ajax.rpc,

    debug: false,
    isMobile: window.innerWidth <= 768
  };
  return { env, menuInfo };
});

//------------------------------------------------------------------------------
// Adapters
//------------------------------------------------------------------------------

function ajaxFetch(route: string, params: any): Promise<any> {
  console.log("RPC", route, params);
  const delay = Math.random() * 150;
  return new Promise(resolve => setTimeout(resolve, delay));
}

/**
 * Load xml templates as a string.
 */
async function loadTemplates(): Promise<string> {
  const result = await fetch("templates.xml");
  if (!result.ok) {
    throw new Error("Error while fetching xml templates");
  }
  return result.text();
}

/**
 * Load all menu items
 */
function loadMenus(): MenuInfo {
  const menuItems: BaseMenuItem[] = (<any>window).odoo.menus;
  if (!menuItems) {
    throw new Error("Cannot find menus description");
  }
  const menuInfo = getMenuInfo(menuItems);
  delete (<any>window).odoo.menus; // overkill?
  return menuInfo;
}

interface BaseMenuItem {
  id: number;
  name: string;
  parent_id: number | false;
  action: string | false;
  icon: string | false;
  children: BaseMenuItem[];
}

/**
 * Generate a valid MenuInfo object from a list of BaseMenuItems. This function
 * is supposed to be called once at startup.
 */
export function getMenuInfo(items: BaseMenuItem[]): MenuInfo {
  const menus: { [key: number]: MenuItem | undefined } = {};
  const actionMap: { [id: number]: MenuItem | undefined } = {};
  const roots: number[] = [];

  // build MenuItems
  for (let root of items) {
    roots.push(root.id);
    addToMap(root);
  }

  function addToMap(m: BaseMenuItem, app?: MenuItem): MenuItem {
    const item: Partial<MenuItem> = {
      id: m.id,
      name: m.name,
      parentId: m.parent_id,
      action: m.action,
      icon: m.icon,
      actionId: -1 // will be filled later on with correct id
    };
    item.app = app || (item as MenuItem);
    item.children = m.children.map(c => addToMap(c, item.app as MenuItem));
    menus[item.id!] = item as MenuItem;
    return item as MenuItem;
  }

  // add proper actionId to every menuitems
  for (let menuId in menus) {
    const menu = menus[menuId]!;
    let menuWithAction = menu && findInTree(menu, m => Boolean(m.action));
    if (menuWithAction) {
      menu.actionId = parseInt(
        (<string>menuWithAction.action).split(",")[1],
        10
      );
      if (menuWithAction === menu) {
        actionMap[menu.actionId] = menu;
      }
    }
  }

  return { menus, roots, actionMap };
}
