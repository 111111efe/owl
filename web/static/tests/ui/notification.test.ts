import { Env, makeEnv } from "../../src/ts/env";
import { Notification as INotification, Store } from "../../src/ts/store/store";
import { Notification } from "../../src/ts/ui/notification";
import * as helpers from "../helpers";

//------------------------------------------------------------------------------
// Setup and helpers
//------------------------------------------------------------------------------

let fixture: HTMLElement;
let store: Store;
let env: Env;
let templates: string;

beforeAll(async () => {
  templates = await helpers.loadTemplates();
});

beforeEach(() => {
  fixture = helpers.makeTestFixture();
  fixture = helpers.makeTestFixture();
  store = helpers.makeTestStore();
  env = makeEnv(store, templates);
});

afterEach(() => {
  fixture.remove();
});

function makeNotification(notif: Partial<INotification> = {}): INotification {
  const defaultNotif = {
    id: 1,
    title: "title",
    message: "message",
    type: "notification",
    sticky: false
  };
  return Object.assign(defaultNotif, notif);
}

//------------------------------------------------------------------------------
// Tests
//------------------------------------------------------------------------------

test("can be rendered", async () => {
  const notif = makeNotification({ title: "title", message: "message" });
  const navbar = new Notification(env, notif);
  await navbar.mount(fixture);
  expect(fixture.innerHTML).toMatchSnapshot();
});

test("can be rendered (type = warning)", async () => {
  const notif = makeNotification({
    title: "title",
    message: "message",
    type: "warning"
  });
  const navbar = new Notification(env, notif);
  await navbar.mount(fixture);
  expect(fixture.innerHTML).toMatchSnapshot();
});

test("can be closed by clicking on it (if sticky)", async () => {
  env.addNotification({
    title: "title",
    message: "message",
    sticky: true
  });

  const navbar = new Notification(env, store.state.notifications[0]);
  await navbar.mount(fixture);
  expect(store.state.notifications.length).toBe(1);
  (<any>fixture.getElementsByClassName("o_close")[0]).click();
  expect(store.state.notifications.length).toBe(0);
});
