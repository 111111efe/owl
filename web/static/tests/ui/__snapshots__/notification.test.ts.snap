// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can be rendered (type = warning) 1`] = `
"<div class=\\"o_notification o_error\\">
    
    <div class=\\"o_notification_title\\">
      <span class=\\"o_icon fa fa-3x fa-exclamation\\" role=\\"img\\" aria-label=\\"Notification title\\" title=\\"Notification title\\"></span>
      title
    </div>
    <div class=\\"o_notification_content\\">
      message
    </div>
  </div>"
`;

exports[`can be rendered 1`] = `
"<div class=\\"o_notification\\">
    
    <div class=\\"o_notification_title\\">
      <span class=\\"o_icon fa fa-3x fa-lightbulb\\" role=\\"img\\" aria-label=\\"Notification title\\" title=\\"Notification title\\"></span>
      title
    </div>
    <div class=\\"o_notification_content\\">
      message
    </div>
  </div>"
`;
